### PR TITLE
Rework console warnings + add defaults + no 3 of same family behaviour + minrankranked and co + maxconnectedgames maxconnectedgamesperuser + fix missing exports + Simple time rethought : added again + Add custom-branches.md + UHMAEAT 2.0 for blitz/live/corr !!! + move to docs + ban "none" + resupport "absolute" + remove ISSA + readme and documentation and warning reworks + move OPTIONS full list -> OPTIONS-LIST.md

### PR DESCRIPTION
**Pull request by wonderingabout**
Many new changes in this PR, some are important :

general look here : https://github.com/wonderingabout/gtp2ogs/tree/simple-time-rethink

easy review here : https://github.com/online-go/gtp2ogs/pull/124/files?utf8=%E2%9C%93&diff=split

- removed console warnings for maxmaintime and co "setting not detected"
- instead, using default values hassle-free, large and friendly for rescue value purposes
- change behaviour if 3 arguments of same family are used : for arguments that have a general+ranked+unranked possible values, new behaviour is now, for example for `--maxmaintime` :  general value (default or manually inputted by bot admin) will be taken into account only if no specific value is used (ranked and/or unranked); else bot admin will have to set up both ranked and unranked values desired and general value will not be taken into account (added notes about that on readme)
- added `--minrankranked` and co : if bot admin wants to specify different minrank and maxrank for ranked and unranked games
- renamed `--maxactivegames` and `--maxtotalgames` to the clearer and more powerful `--maxconnectedgames` and `--maxconnectedgamesperuser` : these names are clearer but also more powerful : in the future, someone may want to add a maxtotalgames per bot and/or per user which is something different (max number of total games in the list on ogs profile, see gnugo page for example : it has a lot of games in the list but gnugo is not connected to almost any of them, bot admins may want to have the option to regulate that in the future) @lemonsqueeze 
- updated readme and ISSA to take all that into account
- added again simple time support as a period time only with `t.per_move`
- added custom-branches.md, see : https://github.com/wonderingabout/gtp2ogs/tree/simple-time-rethink#custom-branches , i'm sure this centralized data will be helpful, we can link to it if someone didnt see it and asks the question
- added UHMAEAT 2.0 for min/max maintime blitz/live/corr +/- ranked/unranked is here !!!
- moved all docs to docs : it's getting quite crowded here on the root directory, similar to what some famous projects did :  https://github.com/django/django and https://github.com/Homebrew/brew for example
- added UHMAEAT 2.0 for period time too
- resupported "absolute" again, but not in defaults (if bot admin wants it, then he need to greatly increase minmaintime*blitz/live above default
- removed ISSA (intuitive semi-syllabic aliases) because we wont input our arguments on command line anymore soon, will use a config file instead
- other preparative and minor changes
- moved the full list of options to OPTIONS-LIST.md
- all options are now clickable links in OPTIONS-LIST.md
- banned is an exception to the rule : we allow both general and ranked/unranked at the same time
- "none" speed is supported again, but removed from defaults

continuation of #123 

continues in #129 , which is on top of this PR

@roy7 @Dorus @windo @lemonsqueeze

### CODE REVIEW

**Comment by wonderingabout** [comment]
just reverted maxmaintime to 7200 default (2 hours)

also added a custom-branches.md note that i'm sure will be helpful, see : https://github.com/wonderingabout/gtp2ogs/tree/simple-time-rethink#custom-branches

@Dorus @roy7

**Comment by wonderingabout** [comment]
TODO : fix as @roy7 reviewed in : https://github.com/online-go/gtp2ogs/pull/123#issuecomment-462088952

**Comment by roy7** [comment]
Maybe getting the config file system together along with the defaults/etc is the way to go. Since you are right a single maintain default for all game speeds would interfere with correspondence.

**Comment by wonderingabout** [comment]
yes, we need to set up separate speed settings for blitz, live, and corr

maxmaintimeall // for all speeds
maxmaintimeblitz
maxmaintimelive
maxmaintimecorr
maxmaintimeblitzranked
etc..
maxperiodtimeblitz
etc...

also indeed we're getting more and more options now, it would be a lot clearer and practical in a config file, so maybe its about time we start working on it

if i knew how to do it, i would definitely would have done it already

but since i cant, i'll just implement blitz live and corr separate settings for now @roy7 @windo @Dorus

**Comment by roy7** [comment]
Yeah one step at a time. :)

**Comment by wonderingabout** [comment]
lool, i really wish we can make it to the master release in time

i want to keep the tradition of a march 2019 master release lol

**Comment by wonderingabout** [comment]
the good thing about it is it will be surprisingly easy now that we have the UHMAEAT

i cant imagine what a pain it would be with the previous code of maintime and periodtime lol

**Comment by wonderingabout** [comment]
so i guess we dont want to merge this until separate settings for blitz, live, corr are implemented

to have effective defaults

**Comment by wonderingabout** [comment]
btw this PR has just been tested to work on beta ogs

**Comment by wonderingabout** [comment]
UHMAEAT 2.0 for min/max maintime blitz/live/corr +/- ranked/unranked is here !!!

big improvements everywhere

i didnt test yet, and didnt do UHMEAT for period times yet :+1: 

but using OOP now :smiley:

**Comment by wonderingabout** [comment]
after testing, current error on UHMAEAT 2.0

with --maxmaintimeblitzranked 120

uhmaeat 2 0

**Comment by wonderingabout** [comment]
edit : actually its just a minor issue of storing the variable 

fischer is dected in time control challenge : 

worksthough

**Comment by wonderingabout** [comment]
first i rebased with the warning first, now gotta debug UHMAEAT 2.0 then commit it again fixed

**Comment by wonderingabout** [comment]
so i found the issue

it's because "blitz" does not exist as a game speed

the only game speed that exist are `t.speed` : 
- "live" for blitz and live,
- "correspondence" for corr

**Comment by wonderingabout** [comment]
also do we still want to reject simple and none if a main time is specified @roy7 ?

`'minmaintimelive', 'Minimum seconds of main time for live AND blitz (rejects time control simple and none)'`

if yes, then i'd need to add it to the UHMAEAT 2.0 again

**Comment by wonderingabout** [comment]
all fixed !! @roy7 @Dorus it works !!!

it all works now !!

works1
works2
works3
works4
works5
works6

**Comment by roy7** [comment]
I don't think we reject a time mode like none or simple because a maintime limit is set. It would apply only when applicable? If someone doesn't want none they should exclude it on the list of allowed time controls. Speaking of which, maybe we should have a default list of allowed time controls and leave none off the list?

**Comment by wonderingabout** [comment]
@roy7 we already do have such a list : 

--timecontrol or --tc Time control(s) to accept (default fischer, byoyomi,simple,canadian,absolute,none)

i think none should be removed from the defaults because it doesnt have a period time support
and absolute too for the same reason 

for example in a blitz absolute game of say 1 minute maintime (currently allowed in defaults), when maintime dries out then the bot timesout

however simple time has a period time support ("time per move"), so we added it again

to sum up, i think absolute and none should be removed from defaults, but simple should be allowed as a period time

in the UHMAEAT, simple time is still allowed

**Comment by wonderingabout** [comment]
btw i finished the UHMAEAT v2 for period times, will commit it soon too :+1:

**Comment by roy7** [comment]
I don't mind leaving 'none' off the default list. Operators can enable it if they really want. Absolute I'm a bit more fuzzy on. Keep in mind none and absolute do need to work properly if the operator turns them on in spite of either other settings. In which case, any min/max maintime set would be the limits on the absolute game timer.

**Comment by wonderingabout** [comment]
i understand

let's consider 2 cases : 

for what is below, remember that maxperiods default is set to 20

### 1) blitz game : 
the current defaults i set for maintime is (can be changed if want)
- minmaintimeblitz : 15 seconds
- maxmaintimeblitz : 5 minutes
then for blitz period time : 
- minperiodtimeblitz : 5 seconds
- maxperiodtimeblitz : 10 seconds (max possible for blitz)

so if user does a blitz absolute game of 15 seconds main time, then if our bot plays 2 seconds per move, we need to remove startup time first, so maybe our bot wont even play one move lol :laughing: 
playing this without a periods number is just crazy lol

### 2) live games

the current defaults i set for maintime is (can be changed if want)
- minmaintimelive : 10 seconds (to continue after blitz)
- maxmaintimelive : 2 hours
then for blitz period time : 
- minperiodtimelive : 10 seconds
- maxperiodtimelive :  2 minutes 

then for absolute, user may play an absolute game of 10 seconds per move
needless to say our bot wont even have time to start, or maybe play one move of honor then lose lol

conclusion : for bots, we just cant allow timecontrols that dont have period time mangement

not only does it break all the balance, but it doesnt even make sense

**Comment by wonderingabout** [comment]
@roy7

**Comment by roy7** [comment]
We certainly can allow absolute even without periods. If an operator wants to allow absolute, they'll set minmaintime to like 5 minutes or something. Or else their bot will lose all of the time until they learn what they are doing. ;)

We can leave none and absolute off the default allowed time modes, I don't mind that. I just don't want them totally banned. Like don't say "We have a minperiods of 1 so all none and absolute games are rejected since those modes don't use periods". Minperiods would only apply to time controls that use periods.

**Comment by wonderingabout** [comment]
ok i see,

yes, removing from the defaults make sense because minmaintimeblitz and minmaintimelive are just way too small to play a game

but you're right that banning them might be a bit extreme

if bot admin wants, he should have the ability to increase minimaintimeblitz to 5 minutes

i'm ok with it

so we ban only none ?

and remove absolute from defaults, while not banning it ?

**Comment by roy7** [comment]
I don't think we ban none either. I think we only remove it from defaults. If a bot operator intentionally wants to allow games with 'none' I think we should let them? 

@TokumotoK what do you think? Should a bot operator be able to run 'none' games if they choose to, or is it better for OGS to simply never allow 'none' as a bot time option in the future?

**Comment by wonderingabout** [comment]
"none" is basically  no time limit !!

allowing this is, i dont even have words to describe it lol

think about it this way

if we want to allow none, then we should just remove all time controls, and make "none" the only time control, everyone will then play at the speed he wants

**Comment by roy7** [comment]
Yes I understand 'none' means no time limits. And bots by default have allowed that ever since bot support was added to OGS. So we should be certain everyone agrees if we want to stop allowing it even by operators who wanted to turn on the none time control for some crazy reason.

Other than the official bots ran by @anoek, there aren't many actual operators out there and I don't think the DarkGo guy will care if none goes away. If @TokumotoK and @Dorus agree then I think we can decide to just kill off 'none' for good.

**Comment by wonderingabout** [comment]
i dont want to become a criminal lol

"killing off" is not needed lol

just ban it and let it wander forever in darkness of dark OGS lol

**Comment by Dorus** [comment]
Just keep none support but do not allow it by default.

Also none games do time out, take like 2-3 weeks. Trust me I've had them expire :)

For other settings, we can just allow time settings per system. So you get

Timing.ranked.canadese.minmain
Timing.unranked.canadese.minmain
Timing.ranked.absolute.minmain
Etc etc. 

If we want to go fancy, you could also add the option for
Timing.minmain 
To fall back too when the specific is not set

**Comment by wonderingabout** [comment]
@Dorus 

the UHMAEAT 2.0 already does that

i'm currently reimplementing "absolute" in the UHMAEAT 2.0, and finishing the exports on UHMAEAT 2.0 maintime to period time

then for "none" i'll still ban it, if someone wants to unban it theres no "support" for none, since by definition its the absence of time lol

or here or not here, "none" doesnt require any support :laughing:

**Comment by wonderingabout** [comment]
tested to work !!!
see commit notes for details !!

@Dorus @roy7 

vorks1
vorks2
vorks3
vorks4
vorks5
vorks6
vorks7

now updating readme

**Comment by wonderingabout** [comment]
@roy7 @Dorus 

i think i finished my work on this PR

tested to work on beta OGS (see screenshots above), merge if it's good for you, else we can review again :1st_place_medal:

**Comment by wonderingabout** [comment]
use this display for easy compare : 

https://github.com/online-go/gtp2ogs/pull/124/files?utf8=%E2%9C%93&diff=split

**Comment by wonderingabout** [comment]
last idea for now : 

grep "this bot is username" from `player_id` rather than `config.username`

todo next time :+1:

File: config.js [hunk 1]
```diff
@@ -210,94 +202,54 @@ console.log("\nYou are using gtp2ogs version 6.0\n- For changelog or latest deve
...
 
-if (argv.maxperiodtime && (argv.maxperiodtimeranked || argv.maxperiodtimeunranked)) {
-    console.log("Warning: You are using --maxperiodtime in combination with --maxperiodtimeranked and/or --maxperiodtimeunranked.\nUse either --maxperiodtime alone, OR --maxperiodtimeranked with --maxperiodtimeunranked.\nBut don't use the 3 maxperiodtime arguments at the same time.");
+if (argv.maxrank && (argv.maxrankranked || argv.maxrankunranked)) {
+    console.log("Warning: You are using --maxrank in combination with --maxrankranked and/or --maxrankunranked. \n Use either --maxrank alone, OR --maxrankranked with --maxrankunranked.\nBut don't use the 3 maxrank arguments at the same time.");
 }
 
-if (argv.minperiodtime && (argv.minperiodtimeranked || argv.minperiodtimeunranked)) {
-    console.log("Warning: You are using --minperiodtime in combination with --minperiodtimeranked and/or --minperiodtimeunranked.\nUse either --minperiodtime alone, OR --minperiodtimeranked with --minperiodtimeunranked.\nBut don't use the 3 minperiodtime arguments at the same time.");
+if (argv.ban && (argv.banranked || argv.banunranked)) {
```
Comment (Dorus) [inline config.js]: Please check this per user, or remove this constraint.

It should be possible to ban some users global, because they're anoying, but other users just for ranked games if they abuse ranked games.

File: config.js [hunk 2]
```diff
@@ -415,7 +371,51 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
...
+                exports.minrankranked = 36 + parseInt(results[1]);
+                exports.proonly = true;
+            } else {
+                console.error("Invalid minrankranked " + argv.minrankranked);
+                process.exit();
+            }
+        } else {
+            console.error("Could not parse minrankranked " + argv.minrankranked);
+            process.exit();
+        }
```
Comment (Dorus) [inline config.js]: These lines are duplicate with 397-415, could be made a function.

File: config.js [hunk 3]
```diff
@@ -436,6 +436,48 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
...
+            } else if (results[2] == "p") {
+                exports.maxrankranked = 36 + parseInt(results[1]);
+            } else {
+                console.error("Invalid maxrankranked " + argv.maxrankranked);
+                process.exit();
+            }
+        } else {
+            console.error("Could not parse maxrankranked " + argv.maxrankranked);
+            process.exit();
+        }
```
Comment (Dorus) [inline config.js]: More duplication.

File: config.js [hunk 4]
```diff
@@ -375,7 +331,7 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
         return false;
     }
 
-    if (argv.ban) {
+    if (argv.ban && !argv.banranked && !argv.banunranked) {
```
Comment (Dorus) [inline config.js]: No need for this check on `!argv.banranked && !argv.banunranked`. just append `--ban` to both lists and append the other two to their own list.

File: config.js [hunk 5]
```diff
@@ -373,7 +329,7 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
...
```
Comment (roy7) [inline config.js]: Make this `if (argv.ban}` so it'll add any global bans to the global ban list. People should be able to use --ban and --banranked and --banunranked. Just because you ban a few people totally doesn't mean you can't also ban someone else only from ranked games.

File: config.js [hunk 6]
```diff
@@ -208,94 +200,54 @@ console.log("\nYou are using gtp2ogs version 6.0\n- For changelog or latest deve
...
 
-if (argv.maxperiodtime && (argv.maxperiodtimeranked || argv.maxperiodtimeunranked)) {
-    console.log("Warning: You are using --maxperiodtime in combination with --maxperiodtimeranked and/or --maxperiodtimeunranked.\nUse either --maxperiodtime alone, OR --maxperiodtimeranked with --maxperiodtimeunranked.\nBut don't use the 3 maxperiodtime arguments at the same time.");
+if (argv.maxrank && (argv.maxrankranked || argv.maxrankunranked)) {
+    console.log("Warning: You are using --maxrank in combination with --maxrankranked and/or --maxrankunranked. \n Use either --maxrank alone, OR --maxrankranked with --maxrankunranked.\nBut don't use the 3 maxrank arguments at the same time.");
 }
 
-if (argv.minperiodtime && (argv.minperiodtimeranked || argv.minperiodtimeunranked)) {
-    console.log("Warning: You are using --minperiodtime in combination with --minperiodtimeranked and/or --minperiodtimeunranked.\nUse either --minperiodtime alone, OR --minperiodtimeranked with --minperiodtimeunranked.\nBut don't use the 3 minperiodtime arguments at the same time.");
+if (argv.ban && (argv.banranked || argv.banunranked)) {
```
Comment (roy7) [inline config.js]: This isn't the same situation as like maxmaintime and maxmaintimeranked etc. Since you might do --ban A,B,C and --banranked D,E. I think we should allow all 3 to be used and just populate the 3 ban lists accordingly.

File: connection.js [hunk 7]
```diff
@@ -271,25 +272,54 @@ class Connection {
...
             conn_log(user.username + " ranking too low: " + humanReadableUserRank + " : min is " + humanReadableMinRank);
             return { reject: true, msg: "Minimum rank is " + humanReadableMinRank + ", your rank is too low." };
         }
 
-        if (user.ranking > config.maxrank) {
+        if ((user.ranking < config.minrankranked) && notification.ranked) {
+            let humanReadableUserRank = rankToString(user.ranking);
+            let humanReadableMinRank = rankToString(config.minrankranked);
+            conn_log(user.username + " ranking too low: " + humanReadableUserRank + " : min for ranked games is " + humanReadableMinRank);
+            return { reject: true, msg: "Minimum rank for ranked games is " + humanReadableMinRank + ", your rank is too low, try unranked game" };
```
Comment (Dorus) [inline connection.js]: The only difference with line 280-284 is the return value here. I got the feeling this could be done with a lot less duplicate code.

The check on line 280 could be `config.minrank && user.ranking < config.minrank || user.ranking && user.ranking < config.minrankranked` and then you only need to return a different message based on `config.minrank` like

```
`${user.username} ranking too low: ${humanReadableUserRank}: min ${config.minrank ? "" : for ranked games} is ${humanReadableMinRank}`
```

btw, consider using string interpolation everywhere.

Ps. I removed another redundant space here. Please keep it clean and remove all spaces before interpunction like `:`, `,`, `.` etc. etc.

File: connection.js [hunk 8]
```diff
@@ -271,25 +272,54 @@ class Connection {
...
+            let humanReadableUserRank = rankToString(user.ranking);
+            let humanReadableMinRank = rankToString(config.minrankunranked);
+            conn_log(user.username + " ranking too low: " + humanReadableUserRank + " : min for ranked games is " + humanReadableMinRank);
+            return { reject: true, msg: "Minimum rank for unranked games is " + humanReadableMinRank + ", your rank is too low" };
+        }
+
+        if ((user.ranking > config.maxrank) && !config.maxrankranked && !config.maxrankunranked) {
             let humanReadableUserRank = rankToString(user.ranking);
             let humanReadableMaxRank = rankToString(config.maxrank);
             conn_log(user.username + " ranking too high: " + humanReadableUserRank + " : max is " + humanReadableMaxRank);
```
Comment (Dorus) [inline connection.js]: Remove the space before `: max is`

File: connection.js [hunk 9]
```diff
@@ -271,25 +272,54 @@ class Connection {
...
+            conn_log(user.username + " ranking too high: " + humanReadableUserRank + " : max for ranked games is " + humanReadableMaxRank);
+            return { reject: true, msg: "Maximum rank for ranked games is " + humanReadableMaxRank + ", your rank is too high, try unranked game" };
+        }
+
+        if ((user.ranking > config.maxrankunranked) && !notification.ranked) {
+            let humanReadableUserRank = rankToString(user.ranking);
+            let humanReadableMaxRank = rankToString(config.maxrank);
+            conn_log(user.username + " ranking too high: " + humanReadableUserRank + " : max for unranked games is " + humanReadableMaxRank);
+            return { reject: true, msg: "Maximum rank for unranked games is " + humanReadableMaxRank + ", your rank is too high" };
+        }
```
Comment (Dorus) [inline connection.js]: Again a block of duplicate code that can be merged together like i explained above.

File: connection.js [hunk 10]
```diff
@@ -305,19 +335,24 @@ class Connection {
...
-        /* "fischer" , "simple", "byoyomi" , "canadian" , "absolute"*/
+        // ** first we eliminate "none" time control :
+        // infinite time is not supportable on OGS bots
+        /* "fischer" , "simple", "byoyomi" , "canadian" , "absolute", "none" */
         let TimeControlString = String(t.time_control);
-        if (TimeControlString === "absolute") {
-            conn_log("Minimum and Maximum main time is not supported in time control " + TimeControlString);
-            return { reject: true, msg: "Period time management is not supported in the time control " + TimeControlString + " , please choose a time control that supports the use of a period time, such as byoyomi,fischer,canadian." };
+        if (TimeControlString === "none") {
+            conn_log("The " + TimeControlString + " time control is impossible to manage for bots on OGS, not allowed : ");
```
Comment (Dorus) [inline connection.js]: We do not get to decide if we allow "none" on ogs. This is up to the bot admin. Please leave this as a config option. (i can accept if the default does not allow none).

File: connection.js [hunk 11]
```diff
@@ -401,145 +436,368 @@ class Connection {
...
+            // now just before TimecontrolString is being tested, we again make sure it has the latest value
+            /*"fischer", "byoyomi", "canadian", "simple", "absolute" */
+            this.TimecontrolString = String(t.time_control);
+            // now just before SpeedString is being tested, we again make sure it has the latest value
+            /* "blitz", "live", "correspondence" */
+            this.SpeedString = String(t.speed);
+
+            // before starting, sanity checks //
+            // sanity check : if not fischer, not byoyomi, not canadian, not simple, not absolute
+            if ((this.TimecontrolString !== "fischer") && (this.TimecontrolString !== "byoyomi") && (this.TimecontrolString !== "canadian") && (this.TimecontrolString !== "simple") && (this.TimecontrolString !== "absolute")) {
```
Comment (Dorus) [inline connection.js]: Why do this check in this incredible complicated way? Before we did `if (!config.allowed_timecontrols[t.time_control]) {` and that's pretty much the same check i think, If anything, write this check like the last one `!["fischer", "byoyomi", "canadian", "simple", "absolute"].includes(this.TimecontrolString)`.

File: connection.js [hunk 12]
```diff
@@ -401,145 +436,368 @@ class Connection {
...
+                conn_log ("error, could not find allowed time control in " + t.time_control);
+                return { reject : true, msg: "error, could not find allowed time control in " + t.timecontrol};
+            }
+            // sanity check : if not "blitz" , not "live" , not "correspondence"
+            if ((this.SpeedString !== "blitz") && (this.SpeedString !== "live") && (this.SpeedString !== "correspondence")) {
```
Comment (Dorus) [inline connection.js]: A) does OGS even allow different speeds?
B) again this check can be greatly simplified.

File: connection.js [hunk 13]
```diff
@@ -401,145 +436,368 @@ class Connection {
...
+                if (config.minmaintimeblitzranked && notification.ranked) {
+                    this.TimeNumber = config.minmaintimeblitzranked;
+                    this.TimeToString = timespanToDisplayString(config.minmaintimeblitzranked);
+                    this.RankedUnrankedGamesIsSentence = "ranked games is ";
+                }
+                if (config.minmaintimeblitzunranked && !notification.ranked) {
+                    this.TimeNumber = config.minmaintimeblitzunranked;
+                    this.TimeToString = timespanToDisplayString(config.minmaintimeblitzunranked);
+                    this.RankedUnrankedGamesIsSentence = "unranked games is ";
+                }
```
Comment (Dorus) [inline connection.js]: 3x duplicated code. I got the feeling this can be simplified.

File: connection.js [hunk 14]
```diff
@@ -401,145 +436,368 @@ class Connection {
...
+            // sanity check : if not "blitz" , not "live" , not "correspondence"
+            if ((this.SpeedString !== "blitz") && (this.SpeedString !== "live") && (this.SpeedString !== "correspondence")) {
+                conn_log ("error, could not find allowed game speed in " + t.speed);
+                return { reject : true, msg: "error, could not find allowed game speed in " + t.speed};
+            }
+            // -> then if sanity checks all pass :       
+
+            //////////// for blitz games : "blitz" //////////////////
+            // min
+            if ((config.minmaintimeblitz || config.minmaintimeblitzranked || config.minmaintimeblitzunranked) && (this.SpeedString === "blitz")) {
```
Comment (Dorus) [inline connection.js]: For sanity, please put the `(this.SpeedString === "blitz")` check first. Makes it more readable to me.

Actually we already check the seperate settings down here, it might be enough to just check fror speed=blitz here. Simplifies the code and the performance is really neglectable.

File: config.js (hunk 5 above)
Comment (wonderingabout) [reply config.js]: ok i understand, these situations are indeed different because ban is a string !

File: config.js (hunk 6 above)
Comment (wonderingabout) [reply config.js]: same !

File: config.js (hunk 1 above)
Comment (wonderingabout) [reply config.js]: yes, same as what i said earlier, ban is the exception that confirms the rule !

File: config.js (hunk 2 above)
Comment (wonderingabout) [reply config.js]: indeed, but i was like let's make this fast

however yes definitely ! will do

File: config.js (hunk 3 above)
Comment (wonderingabout) [reply config.js]: same !

File: config.js (hunk 4 above)
Comment (wonderingabout) [reply config.js]: same !

File: connection.js (hunk 7 above)
Comment (wonderingabout) [reply connection.js]: this is quite complicated @Dorus :rofl:

Comment (wonderingabout) [reply connection.js]: oh i see you mean to use `blablah ${1} blablah` instead of "blablah " + v1 + " blablah";

i see but i did it to keep in sync with the current code
but also because on gedit the "blablah" is pink so easy to read, but the `blablah` is dark same color as plain text, hard to read

Comment (wonderingabout) [reply connection.js]: also if i start changing this then i would have to change all the other ones for consistency :sob: 

but yes it's better i think ! ty for information !

File: connection.js (hunk 9 above)
Comment (wonderingabout) [reply connection.js]: will try to write a function :+1: 

indeed

File: connection.js (hunk 8 above)
Comment (wonderingabout) [reply connection.js]: will try to write a function :+1:

Comment (wonderingabout) [reply connection.js]: this can indeed save a lot of code maintainance and increase a lot code efficiency

File: connection.js (hunk 10 above)
Comment (wonderingabout) [reply connection.js]: ok, will just comment out this code then (maybe we want to change this policy in the future)

File: connection.js (hunk 11 above)
Comment (wonderingabout) [reply connection.js]: because i didnt know about that yet lol, but indeed it's smaller :laughing: 

however seeing how big the UHMAEAT is, this check is really needed, and it helped in ebugging already, so we'd rather keep it :+1:

File: connection.js (hunk 12 above)
Comment (wonderingabout) [reply connection.js]: A) yes ofc
similar to how humans can play different speeds
it didnt exist in the past on ogs, but now we have it

simple example : 
the old --maxmaintime 7200 (2 hours) would not allow to play corr games (time too short for corr, but now we can do --maxmaintimelive 7200 --maxmaintimecorr 609800 (2 hours, 7 days)

B) indeed, will try !!

File: connection.js (hunk 13 above)
Comment (wonderingabout) [reply connection.js]: i can try to write a function, but for periodtime canadian it's not the same (it uses a different formula)

File: connection.js (hunk 14 above)
Comment (wonderingabout) [reply connection.js]: actually i did it for the exact same reason : for performance

i thought if this check fails then we dont waste cpu power to check all the checks below !

and for blitz yes it is clearer, similar to what i did in #121

**Comment by wonderingabout** [comment]
ok i'll try to write a few functions, especially for the UHMAEAT !!!

then i'll add --boardsizeranked --boardsizeunranked
example : if botadmin wants to play 19x19 in ranked only, but wants to also allow 13x13 and 9x9 in unranked because his bot is less good at it

File: connection.js (hunk 7 above)
Comment (Dorus) [reply connection.js]: Ah yes i got the same color problem in my editor. Really need to look into a way to fix that first i think :(

File: connection.js (hunk 13 above)
Comment (Dorus) [reply connection.js]: oh for different formula, it might not be a function after all. Or possibly a smaller part can be made into a function?

Comment (wonderingabout) [reply connection.js]: we can see the specific problem of canadian period time later, (only adds 3 more variables : 
- nc = n * stones;
- ncToString = timespanToDisplayString(nc);
- ncSentence = some small blablah

Comment (wonderingabout) [reply connection.js]: but gonna try to UHMAEAT 3.0

it would indeed remove a lot of code maintainance

**Comment by wonderingabout** [comment]
note for myself later : 

conn_log message of UHMAEAT 2.0 for period times, in canadian (conn_log message needs to be rewritten), and also some "undefined" needs to be clearer

**Comment by anoek** [comment]
@wonderingabout In the future please distill your pull requests into smaller digestible chunks, that way they can be reviewed and merged individually. 

Also the title on this PR is unreadable. Please keep a title to a few words, under 80 characters at most if possible. Put expanded details in the body of the commit message.

**Comment by wonderingabout** [comment]
@anoek @roy7 will try to improve it next time, indeed i lack some experience in committing, i hope it will be better in the future

**Comment by wonderingabout** [comment]
i'm finishing some commits, then will remove this last commit and continue on a new PR

TODO : fix UHMAEAT v2 canadian periodtime reject message

**Comment by wonderingabout** [comment]
@Dorus 

you may want to "textify" this is a more human readable way (extract in `9,13,19,etc` numbers one by one, and print 9x9 13x13 19x19)

in this code 

```
        if (!config.allowed_sizes[notification.width] && !config.allow_all_sizes && !config.allow_custom_sizes && !config.boardsizeranked && !config.boardsizeunranked) {
            conn_log("square board size " + notification.width + "x" + notification.height + " is not an allowed size, rejecting challenge");
            return { reject: true, msg: "Board size " + notification.width + "x" + notification.height + " is not allowed, please choose one of the allowed square board sizes (same width and height, for example if allowed boardsizes are 9,13,19, it means you can play only 9x9 , 13x13, and 19x19), these are the allowed square board sizes : " + config.boardsize };
        }
```

will be moved in a new PR, but notifiying for now already

**Comment by wonderingabout** [comment]
@Dorus 

new reject message would then be 

```
"Board size " + notification.width + "x" + notification.height + " is not allowed, please choose one of the allowed square board sizes : " + n1 + "x" + n1 + "," +  n2 + "x" + n2 + "," + n3 + "x" + n3 + "," + etc + "x" + etc;
```

**Comment by wonderingabout** [comment]
continues in #129

**Comment by wonderingabout** [comment]
@roy7 @Dorus 

ready to merge again,

read #129 for remaining issues and new features

File: config.js [hunk 15]
```diff
@@ -415,7 +375,51 @@ if (argv.botid || argv.bot || argv.id || argv.minrankedhandicap || argv.maxranke
         }
     }
 
-    if (argv.maxrank) {
+    if (argv.minrankranked) {
+        let re = /(\d+)([kdp])/;
+        let results = argv.minrank.toLowerCase().match(re);
```
Comment (roy7) [inline config.js]: Shouldn't this be argv.minrankranked?

File: connection.js [hunk 16]
```diff
@@ -37,7 +37,8 @@ class Connection {
         });
 
         this.connected_games = {};
-        this.games_by_player = {};     // Keep track of active games per player
+        this.connected_game_timeouts = {};
```
Comment (roy7) [inline connection.js]: I think you need to do a git pull off of devel. `this.connected_game_timeouts = {};` was removed in a merged PR, it shouldn't be getting merged back in.

File: config.js [hunk 17]
```diff
@@ -112,11 +48,14 @@ exports.updateFromArgv = function() {
...
         .describe('corrqueue', 'Process correspondence games one at a time')
-        .describe('maxtotalgames', 'Maximum number of total games')
-        // maxtotalgames is actually the maximum total number of connected games for your bot 
-        // which means the maximum number of games your bot can play at the same time (choose a low number to regulate your GPU use)
+        .describe('maxconnectedgames', 'Maximum number of connected games for all users')
+        .default('maxconnectedgames', 20)
+        // maxconnectedgames is actually the maximum number of connected games for all users 
+        // against your bot, which means the maximum number of games your bot can play at the same time 
+        // (choose a low number to regulate your computer performance and stability)
         // (correspondence games are currently included in the total connected games count if you use `--persist` )
```
Comment (roy7) [inline config.js]: I don't believe --persist matters, right? Connected game list is every game you are a part of on OGS. Maybe change to:

> (correspondence games are currently included in the total connected games count)

Comment (wonderingabout) [reply config.js]: indeed yes
weird i'm sure i had removed this line some time ago

File: connection.js (hunk 16 above)
Comment (wonderingabout) [reply connection.js]: i pulled upstream devel and manually resolved merge conflict in https://github.com/online-go/gtp2ogs/pull/124/commits/42293545ca93fb513ed998026bd2366fd508f8cd @roy7

Comment (roy7) [reply connection.js]: Ok just remove that line from your copy then please. There is no more connected_game_timeouts array in current /devel.

Comment (wonderingabout) [reply connection.js]: i'm still unsure if i pulled correctly

so i'd rather have a look at it next time @roy7 

maybe its not the only difference with latest devel
